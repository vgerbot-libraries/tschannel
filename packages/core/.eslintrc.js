module.exports = {
    parser: '@typescript-eslint/parser',
    parserOptions: {
        ecmaVersion: 2020,
        sourceType: 'module',
        ecmaFeatures: {
            jsx: true,
            impliedStrict: true
        },
        project: ['./tsconfig.json', 'test/tsconfig.json'],
        tsconfigRootDir: __dirname
    },
    extends: [
        'plugin:@typescript-eslint/recommended', //  Use the recommended rules from the @typescript-eslint/eslint-plugin
        'prettier', // Use eslint-config-prettier to disable ESLint rules from @typescript-eslint/eslint-plugin that would conflict with prettier
        'plugin:prettier/recommended' // Enables eslint-plugin-prettier and eslint-config-prettier
    ],
    rules: {
        'prettier/prettier': 'error',
        'no-var-require': 'off',
        'no-console': 'off',
        'no-bitwise': 'off',
        quotes: ['error', 'single'],
        'max-len': ['error', 120],
        '@typescript-eslint/no-inferrable-types': 'off',
        '@typescript-eslint/explicit-module-boundary-types': 'off',
        '@typescript-eslint/ban-types': 'off',
        'arrow-parens': 'off'
    }
};
