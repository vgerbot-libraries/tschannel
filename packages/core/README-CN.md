# @vgerbot/channel ![tschannel workflow](https://github.com/y1j2x34/channel-ts/actions/workflows/runtest.yml/badge.svg) [![codecov](https://codecov.io/gh/vgerbot-libraries/tschannel/branch/master/graph/badge.svg?token=fPomBmOknB)](https://codecov.io/gh/vgerbot-libraries/tschannel)

---

这是一个由 Typescript 实现的消息传递抽象层，他的目的是封装消息传递的细节，让 js 能够构造存在于不同上下文的类以及像调用普通异步方法一般调用不同上下文的方法。

## 功能

1. 封装：封装通信细节，提供一致上层API
2. 隔离：数据被安全地隔离在不同的 channel-id 中。
3. API：惯用的API，远程调用和普通异步调用一样简单，支持回调函数，异常处理和普通异步方法一致
4. 并行：支持分解任务到多个目标上下文并行执行
5. 扩展：支持自定义 communicator 以支持更多通信方式

## 开始使用

### 安装依赖和环境搭建

```sh
npm i -s @vgerbot/channel
```

为了能够更方便地使用 channel API, 在开始之前， 先安装 [`@vgerbot/channel-transformer`](https://www.npmjs.com/package/@vgerbot/channel-transformer), 其具体如何配置方法可以npm或者github仓库上的 README 文档说明。

```sh
npm i -D @vgerbot/channel-transformer
```

transformer 用法参考 <../packages/transformer/README.md>

### 示例及用法

// api.ts
```ts
export interface SpellChecker {
    saveToDictionary(word: string);
    setCaseSensitive(caseSensitive: boolean);
    check(sentence: string): boolean;
}
```

task.ts
```ts
import { channel } from '@vgerbot/channel';
import { SpellChecker } form './api';

const chnl = channel('worker-channel')
    .connectToMainThread()
    .create();

chnl.def_method(function performCPUIntensiveCalculation() {
    return 'Result!';
});

chnl.def_class(class DefaultSpellCheckerImpl implements SpellChecker {
    saveToDictionary(word: string) {}
    setCaseSensitive(caseSensitive: boolean) {}
    check(sentence) {
        return true;
    }
})
```

```ts
import { channel } from '@vgerbot/channel';
import { SpellChecker } form './api';

const chnl = channel('worker-channel')
    .connectToWorker('./task.js')
    .create();

const performCPUIntensiveCalculation = chnl.get_method('performCPUIntensiveCalculation');
performCPUIntensiveCalculation().then(console.log) // Console Output: "Result!"

const DefaultSpellCheckerImpl = chnl.get_class<SpellChecker>('DefaultSpellCheckerImpl');

const spellChecker = new DefaultSpellCheckerImpl();

spellChecker.saveToDictionary('halo');
spellChecker.setCaseSensitive(false);
spellChecker.check('Halo world!').then(console.log); // Console Output: true

spellChecker.__destroy__(); // 远程实例无法被GC清除，需要手动销毁。
```

更多示例参考 <./packages/examples> 和单元测试

## 许可证

`@vgerbot/channel` 库是根据 MIT 许可证的条款发布的。
