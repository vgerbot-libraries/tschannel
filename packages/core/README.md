# @vgerbot/channel ![tschannel workflow](https://github.com/y1j2x34/channel-ts/actions/workflows/runtest.yml/badge.svg) [![codecov](https://codecov.io/gh/vgerbot-libraries/tschannel/branch/master/graph/badge.svg?token=fPomBmOknB)](https://codecov.io/gh/vgerbot-libraries/tschannel)

