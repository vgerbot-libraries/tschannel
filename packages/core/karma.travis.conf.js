'use strict';
const path = require('path');

const plugins = require('./build/rollup.plugins');
const rollupPluginIstanbul = require('rollup-plugin-istanbul');
const baseConfig = require('./karma.base.conf');
const pkg = require('./package.json');
const channelTsTransformer = require('@vgerbot/channel-transformer').channelTransformerFactory;

module.exports = function (config) {
    const coverageIstanbulReporter = {
        reports: ['html', 'text-summary', 'cobertura'],
        dir: path.join(__dirname, 'coverage'),
        skipFilesWithNoCoverage: true
    };

    const rollupPlugins = [
        plugins.typescript({
            tsconfig: 'test/tsconfig.json',
            transformers: [
                (languageService) => {
                    const program = languageService.getProgram();
                    return {
                        before: [channelTsTransformer(program)]
                    };
                }
            ]
        }),
        plugins.nodeResolve(),
        plugins.commonjs({
            include: /node_modules/,
            ignore: ['js-base64'],
            sourceMap: false,
            namedExports: {
                chai: ['expect']
            }
        }),
        rollupPluginIstanbul({
            exclude: ['test/**/*.ts', '**/node_modules/**/*'],
            instrumenterConfig: {
                embedSource: false,
                debug: false
            }
        }),
        plugins.printError()
    ];
    config.set(Object.assign({}, baseConfig, {
        preprocessors: {
            'test/**/*.ts': ['rollup']
        },
        files: baseConfig.files.concat(
            'test/utils/coverage-utils.ts'
        ),
        rollupPreprocessor: {
            context: 'this',
            watch: false,
            output: {
                format: 'iife',
                name: pkg.library,
                sourcemap: false
            },
            plugins: rollupPlugins,
            onwarn: function(warning) {
                if (warning.code === 'CIRCULAR_DEPENDENCY') {
                    return;
                }
                // console.warn(`(!) ${warning.message}`);
            }
        },
        reporters: ['mocha', 'coverage-istanbul'],
        coverageIstanbulReporter: coverageIstanbulReporter,

        pingTimeout: 1000 * 3000,
        browserNoActivityTimeout: 1000 * 300,

        logLevel: config.LOG_ERROR,

        singleRun: true,

        plugins: baseConfig.plugins.concat([
            'karma-coverage-istanbul-reporter'
        ])
    }));
};
