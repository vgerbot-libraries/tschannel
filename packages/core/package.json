{
    "name": "@vgerbot/channel",
    "version": "1.0.0",
    "description": "A TypeScript-based cross-context messaging library",
    "main": "dist/index.cjs.js",
    "module": "dist/index.esm.js",
    "browser": "dist/index.umd.js",
    "typings": "./dist/index.d.ts",
    "distDir": "dist/",
    "library": "ChannelTS",
    "engines": {
        "node": ">=8.0"
    },
    "directories": {
        "src": "src",
        "test": "test"
    },
    "scripts": {
        "release": "standard-version",
        "lint": " eslint {src,test}/**/*.{js,ts,tsx}",
        "format": "eslint --fix {src,test}/**/*.{js,ts,tsx}",
        "pretest:debug": "npm run lint",
        "debug": "node --inspect ./node_modules/.bin/karma start ./karma.debug.conf.js --browsers Chrome",
        "test": "karma start ./karma.travis.conf.js --browsers HeadlessChrome --single-run true",
        "debug:ci": "node --inspect ./node_modules/.bin/karma start ./karma.travis.conf.js --browsers Chrome --single-run false",
        "codecov": "codecov -f ./coverage/cobertura-coverage.xml --flags '@vgerbot/channel'",
        "test:chromeheadless": "karma start ./karma.debug.conf.js --browsers HeadlessChrome --single-run true",
        "build": "rollup -c ./rollup.config.js --environment 'production'",
        "gendoc": "typedoc ./src  --target es5 --out dist/docs --name channel-ts --hideGenerator",
        "debug:transformer": "node --inspect node_modules/.bin/rollup -c ./rollup-debug.config.js --watch -i test/specs/WebWorkerCommunicator/index.spec.ts -o dist/test.js"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/y1j2x34/channel-ts.git"
    },
    "bugs": {
        "url": "https://github.com/y1j2x34/channel-ts/issues"
    },
    "keywords": [
        "RMI",
        "channel",
        "remote",
        "method",
        "invocation",
        "iframe",
        "websocket",
        "webworker"
    ],
    "files": [
        "dist"
    ],
    "homepage": "https://github.com/y1j2x34/channel-ts/blob/master/packages/channel-ts#readme",
    "author": "'y1j2x34' <y1j2x34@qq.com> (http://www.vgerbot.com/)",
    "license": "ISC",
    "devDependencies": {
        "@rollup/plugin-commonjs": "22.0.2",
        "@rollup/plugin-json": "4.1.0",
        "@rollup/plugin-node-resolve": "14.1.0",
        "@rollup/plugin-strip": "2.1.0",
        "@types/chai": "^4.2.7",
        "@types/chai-as-promised": "^7.1.3",
        "@types/istanbul-lib-coverage": "^2.0.3",
        "@types/karma": "^6.3.3",
        "@types/mocha": "^9.1.1",
        "@types/node": "^18.7.18",
        "@types/sinon": "^10.0.13",
        "@types/sinon-chai": "^3.2.5",
        "@types/websocket": "1.0.5",
        "@typescript-eslint/eslint-plugin": "^5.37.0",
        "@typescript-eslint/parser": "^5.37.0",
        "@vgerbot/karma-rollup-preprocessor": "^7.0.2",
        "chai": "^4.3.6",
        "chai-as-promised": "^7.1.1",
        "@vgerbot/channel-transformer": "^1.0.0",
        "codecov": "*",
        "eslint": "^8.22.0",
        "eslint-config-prettier": "^8.5.0",
        "eslint-plugin-prettier": "^4.2.1",
        "fs-extra": "10.1.0",
        "istanbul-lib-coverage": "^3.0.0",
        "karma": "6.4.0",
        "karma-chai": "^0.1.0",
        "karma-chrome-launcher": "3.1.1",
        "karma-coverage-istanbul-reporter": "3.0.3",
        "karma-mocha": "^2.0.1",
        "karma-mocha-reporter": "^2.2.5",
        "karma-rollup-preprocessor": "7.0.8",
        "karma-sinon": "^1.0.5",
        "karma-sinon-chai": "^2.0.2",
        "karma-sourcemap-loader": "0.3.8",
        "mocha": "^10.0.0",
        "phantomjs-polyfill-object-assign": "0.0.2",
        "prettier": "2.7.1",
        "rollup": "2.79.0",
        "rollup-plugin-istanbul": "3.0.0",
        "rollup-plugin-node-builtins": "2.1.2",
        "rollup-plugin-typescript2": "0.34.0",
        "rollup-plugin-uglify": "6.0.4",
        "sinon": "^14.0.0",
        "sinon-chai": "^3.7.0",
        "ts-node": "^10.9.1",
        "typedoc": "0.23.14",
        "typescript": "4.8.3"
    },
    "dependencies": {
        "tslib": "^2.2.0",
        "txon": "^1.0.1"
    }
}
