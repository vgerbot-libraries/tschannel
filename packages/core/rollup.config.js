const plugins = require('./build/rollup.plugins');
const pkg = require('./package.json');

function createOutputConfig(file, format, cfg = {}) {
    return Object.assign(
        {
            file,
            format,
            sourcemap: true,
            name: pkg.library,
            exports: 'named'
        },
        cfg || {}
    );
}

module.exports = [
    [pkg.browser, 'umd'],
    [pkg.module, 'es'],
    [pkg.main, 'cjs']
].map(confs => createOutputConfig(...confs))
    .map(output => {
        console.log(output);
        return {
            input: 'src/index.ts',
            output,
            plugins: [
                plugins.typescript({
                    tsconfigOverride: {
                        compilerOptions: {
                            target: output.format === 'es' ? 'ESNext' : 'ES5'
                        }
                    }
                }),
                plugins.nodeResolve({
                    mainFields: ['main', 'browser', 'module']
                }),
                plugins.commonjs({
                    include: /node_modules/,
                    ignore: ['js-base64'],
                    sourceMap: false
                }),
                plugins.printError()
            ],
            external: ['txon']
        }
    })
