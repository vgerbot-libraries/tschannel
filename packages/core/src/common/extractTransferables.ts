export function extractTransferables(data: unknown): Transferable[] {
    if (!data) {
        return [];
    }

    const transferables: Transferable[] = [];

    function findTransferables(value: unknown, visited: Set<unknown>) {
        if (visited.has(value)) {
            return;
        }
        visited.add(value);
        if (
            value instanceof ArrayBuffer ||
            value instanceof MessagePort ||
            value instanceof ImageBitmap ||
            value instanceof OffscreenCanvas
        ) {
            transferables.push(value);
        } else if (Array.isArray(value)) {
            value.forEach(it => findTransferables(it, visited));
        } else if (value && typeof value === 'object') {
            for (const key in value) {
                findTransferables((value as Record<string, unknown>)[key], visited);
            }
        }
    }

    findTransferables(data, new Set());

    return transferables;
}
