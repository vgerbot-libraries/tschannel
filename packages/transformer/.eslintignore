commitlint.config.js
jest.config.js
{dist,test_dist}/**/*.js
test/fixtures
test/mock
node_modules
