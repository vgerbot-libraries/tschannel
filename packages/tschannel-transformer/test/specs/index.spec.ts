import { transpile } from '../common/compiler';
import transformer from '../../src';

describe('@tschannel/transformer', () => {
    it('Should run normally with the transformer', async () => {
        const source = `
            import { Channel } from '@tschannel/core';
            export let x: string = 'string';
            console.info(Channel);
        `;
        const output = transpile(source, {
            channelTransformer: transformer
        });
        expect(output).toMatchSnapshot();
    });
    it('Should transform the interface parameter correctly', () => {
        const source = `
            import { Channel } from '@tschannel/core';
            const channel = new Channel();
            channel.rclass<RemoteAPI>();
            interface RemoteAPI {
                method(){}
            }
        `;
        const output = transpile(source, {
            channelTransformer: transformer
        });
        expect(output).toMatchSnapshot();
    });
});
