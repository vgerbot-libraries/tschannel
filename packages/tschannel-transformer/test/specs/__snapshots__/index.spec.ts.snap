// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@tschannel/transformer Should run normally with the transformer 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.x = void 0;
const core_1 = require(\\"@tschannel/core\\");
exports.x = 'string';
console.info(core_1.Channel);
"
`;

exports[`@tschannel/transformer Should transform the interface parameter correctly 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const core_1 = require(\\"@tschannel/core\\");
const channel = new core_1.Channel();
channel.rclass(\\"RemoteAPI\\", class RemoteAPIImpl_1 {
    method() { }
});
{ }
"
`;
